src/object.d

```diff
@@ -111,18 +111,34 @@ int __cmp(T)(scope const T[] lhs, scope const T[] rhs) @trusted
     }
 }
 
-// Compare class and interface objects for ordering.
-private int __cmp(Obj)(Obj lhs, Obj rhs)
-if (is(Obj : Object))
+// Compare class objects for ordering.
+int __cmp(C1, C2)(C1 lhs, C2 rhs)
+if (((is(C1 == class) || is(C1 == typeof(null))) && is(C1 : Object)) &&
+    ((is(C2 == class) || is(C2 == typeof(null))) && is(C2 : Object)))
```

Comment:
I think it allows for a more optimal implementation if `C1` or `C2` is explicitly `null`.  I also think the function can be evaluated at compile-time if one of the arguments is explicitly `null`.  That's what it looks like to me, but I could be wrong.

Reply:
If what I said above is true, it would be nice to add some `unittest` code to demonstrate that and ensure it remains that way now and into the future.

Reply:
But the check for both `class` and `Object` looks redundant to me.

Reply:
Comparing C++ classes is different to comparing D classes.

Reply:
I know, that’s what I said in the beginning. But shouldn’t it be enough to check for inheriting from `Object`?

Reply:
It seems so.
https://run.dlang.io/is/e6XTX5
```d
import std.stdio;

extern(C++) class CppClass
{
    int foo() { return 0; }
}

void main()
{
    writeln(is(typeof(null) : Object));
    writeln(is(typeof(CppClass.init) : Object));
}
```
Output:
```
true
false
```

Reply:
The right signature should be:

```d
int __cmp(C1, C2)(C1 lhs, C2 rhs)
if (is(C1 : const(Object)) && is(C2 : const(Object)))
```

i.e. we're trying to capture all D objects with immutable and const qualifiers

Reply:
@edi33416 I think you also want to capture interfaces, so:

```d
int __cmp(C1, C2)(C1 lhs, C2 rhs)
if ((is(C1 : const(Object)) || is(C1 == interface)) && (is(C2 : const(Object)) || is(C2 == interface)))
```

Reply:
Or does the compiler insert the casts from interface to Object?

Reply:
Best setup is to capture interfaces at library level though.

Reply:
> Or does the compiler insert the casts from interface to Object?

No. Because an interface and be a COM interface.

Reply:
> if (is(C1 : const(Object)) && is(C2 : const(Object)))

I would add `const` to the parameters instead of the constraints.

Reply:
@jacob-carlborg we don't want to restrict comparison to constant objects. Per https://run.dlang.io/is/iTtY7T, it seems the simplest comprehensive sig is:

```d
int __cmp(C1, C2)(C1 lhs, C2 rhs)
if (is(C1 == class) || is(C1 == interface) || is(C1 : const typeof(null))
        &&
    is(C2 == class) || is(C2 == interface) || is(C2 : const typeof(null)));
```

Reply:
@edi33416 later as we introduce ProtoObject we can change the sig appropriately

Reply:
Also: `__cmp` should return `auto`, not `int` so as to accommodate classes that define `opCmp` to return e.g. `float`.

Reply:
> we don't want to restrict comparison to constant objects.

You can pass anything to a `const` parameter, mutable, immutable and const. But of course `opCmp` would need to be `const`. Is that the problem?

Reply:
> ```d
> int __cmp(C1, C2)(C1 lhs, C2 rhs)
> if (is(C1 == class) || is(C1 == interface) || is(C1 : const typeof(null))
>         &&
>     is(C2 == class) || is(C2 == interface) || is(C2 : const typeof(null)));
> ```

The above constraint will include C++ and Objective-C classes and interfaces and COM interfaces. Is that intentional? In that case there needs to be tests for all of those.